firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/send/ReportQueue.java

```diff
@@ -0,0 +1,213 @@
+      OnDemandCounter onDemandCounter) {
+    this.ratePerMinute = ratePerMinute;
+    this.base = base;
+    this.stepDurationMs = stepDurationMs;
+    this.transport = transport;
+    this.onDemandCounter = onDemandCounter;
+
+    // The queue capacity is the per-minute rate number. // TODO(mrober): Round up to next int?
+    queueCapacity = (int) ratePerMinute;
+    queue = new ArrayBlockingQueue<>(queueCapacity);
+    singleThreadExecutor = new ThreadPoolExecutor(1, 1, 0L, TimeUnit.MILLISECONDS, queue);
+
+    step = 0;
+    lastUpdatedMs = 0;
+  }
+
+  /**
+   * Enqueue a report to send through Google DataTransport. If the queue is full, drop the report.
+   *
+   * <p>The report will be sent according to the per-minute rate. avoiding bursts.
+   */
+  TaskCompletionSource<CrashlyticsReportWithSessionId> enqueueReport(
+      CrashlyticsReportWithSessionId reportWithSessionId, boolean isOnDemand) {
+    synchronized (queue) {
+      TaskCompletionSource<CrashlyticsReportWithSessionId> tcs = new TaskCompletionSource<>();
+      if (isOnDemand) {
+        onDemandCounter.incrementRecordedOnDemandExceptions();
+        if (isQueueAvailable()) {
+          Logger.getLogger().d("Enqueueing report: " + reportWithSessionId.getSessionId());
+          Logger.getLogger().d("Queue size: " + queue.size());
+          singleThreadExecutor.execute(new ReportRunnable(reportWithSessionId, tcs));
+
+          // TODO(mrober): Avoid this, so queued tasks can still fail properly.
+          // Complete the task right away to not block on-demand callers.
+          Logger.getLogger().d("Closing task for report: " + reportWithSessionId.getSessionId());
+          tcs.trySetResult(reportWithSessionId);
+
+          return tcs;
+        }
+
+        calcStep();
+        Logger.getLogger()
+            .d("Dropping report due to queue being full: " + reportWithSessionId.getSessionId());
+        onDemandCounter.incrementDroppedOnDemandExceptions();
+        tcs.trySetResult(reportWithSessionId); // Complete the task right away.
+        return tcs;
+      }
+
+      sendReport(reportWithSessionId, tcs);
+      return tcs;
+    }
+  }
+
+  /** Send the report to Crashlytics through Google DataTransport. */
+  private void sendReport(
+      CrashlyticsReportWithSessionId reportWithSessionId,
+      TaskCompletionSource<CrashlyticsReportWithSessionId> tcs) {
+    Logger.getLogger()
+        .d("Sending report through Google DataTransport: " + reportWithSessionId.getSessionId());
+    transport.schedule(
+        Event.ofUrgent(reportWithSessionId.getReport()),
+        error -> {
+          if (error != null) {
+            tcs.trySetException(error);
+            return;
+          }
+          tcs.trySetResult(reportWithSessionId);
+        });
+  }
+
+  private boolean isQueueAvailable() {
+    return queue.size() < queueCapacity;
+  }
+
+  private boolean isQueueFull() {
+    // TODO(mrober): Remove this after testing.
```

Comment:
Is this ready to be removed? If not, can you add a little more context to the TODO?

Reply:
Done